### PR TITLE
First implementation of TruncatedNormal distribution

### PR DESCRIPTION
**Pull request by snexus**
Following discussion in https://discourse.pymc.io/t/truncated-normal-distribution-with-optimal-truncation-thresholds-via-densitydist/1396/9, here is initial implementation of TruncatedNormal distribution.

### CODE REVIEW

**Comment by fonnesbeck** [comment]
Why is this preferable to using `Bound`, which is [the current recommended approach](https://docs.pymc.io/api/bounds.html) for truncating arbitrary continuous variables?

File: pymc3/distributions/continuous.py [hunk 1]
```diff
@@ -434,6 +434,202 @@ def _repr_latex_(self, name=None, dist=None):
...
+    """
+
+    def __init__(self, mu=0, sd=None, tau=None, a=None, b=None, **kwargs):
+        tau, sd = get_tau_sd(tau=tau, sd=sd)
+        self.sd = tt.as_tensor_variable(sd)
+        self.tau = tt.as_tensor_variable(tau)
+        self.a = tt.as_tensor_variable(a) if a is not None else a
+        self.b = tt.as_tensor_variable(b) if b is not None else b
+        self.mu =  tt.as_tensor_variable(mu)
+        self.mean = self.mu
```
Comment (ColCarroll) [inline pymc3/distributions/continuous.py]: this isn't the mean of a truncated normal, and may cause problems when it is used as an initial point

**Review by ColCarroll — COMMENTED** [review]
Looks like a good start! Still need to add tests to `test_distributions` and `test_distributions_random`

**Comment by ColCarroll** [comment]
`Bound` uses rejection sampling to generate samples, while `scipy.truncnorm` uses an inverse CDF (I think? it is quite fast even for silly cases like `truncnorm(mu=100, sd=1, a=0, b=1)`).

**Comment by junpenglao** [comment]
Also bounded variable cannot assign observation.

File: pymc3/distributions/continuous.py (hunk 1 above)
Comment (snexus) [reply pymc3/distributions/continuous.py]: Sure, no problem, wasn't sure where it is used. Do we need also proper implementations of median and mode?

**Comment by snexus** [comment]
Also, as I understand bounding doesn't result in proper distribution since it doesn't integrate to 1. TruncatedNormal has proper normalization...

File: pymc3/distributions/continuous.py (hunk 1 above)
Comment (ColCarroll) [reply pymc3/distributions/continuous.py]: looks good! no need to implement the others (unless it is trivial).  it isn't documented too well, but each distribution has `defaults` defined as a list, and iterates through to find one that has finite log probability to initialize with.  I think the "default" for `defaults` is `[mean, median, mode]`.

**Comment by ColCarroll** [comment]
Looks almost ready to go: just needs a test in `test_distributions_random.py`.  The logic in `.random` recently got revamped, so if it doesn't pass cleanly please push anyways and I can try to help debug.

**Comment by snexus** [comment]
Done. Hope test case is generated correctly for TestScalarParameterSamples...

**Comment by ColCarroll** [comment]
This looks great! Sorry to keep piling on requests, but I realized that this also needs a line in `RELEASE_NOTES.md`

**Comment by ColCarroll** [comment]
Thanks much for the contribution!

**Comment by snexus** [comment]
Thank you! Happy to help!

**Comment by gBokiau** [comment]
Hey this is nice! 

The code crawl 😬though. Not that I'm keeping count, but this could well be the 5th implementation of scaled squared euclidian distance in the code base.

So, a next logical step would be to create a `Truncatable` class that adds `lower`/`upper` parameters to dists that implement it (so long as they have a `cdf()` method). Truncate all the things!

In the spirit of that, I would suggest renaming `a` and `b`, to `lower` and `upper`, to avoid conflicts, e.g. in the case of truncated beta distributions.

**Comment by junpenglao** [comment]
@gBokiau that's a great suggestion. 
Also we need transformation for bounded variables, something similar to:
https://github.com/pymc-devs/pymc3/blob/master/pymc3/distributions/bound.py#L137-L151
A `Truncatable` class should solve this nicely.

**Comment by gBokiau** [comment]
@junpenglao right, like, fall back on `Bound` method if there's no `cdf` method? And maybe issue a warning. NB I don't like the bound method much, as it in effect scales the logp, and thus artificially decreases the weight of bounded dists in the model — so a warning might be warranted. (edit: I see this was part of the discussion on discourse) (edit 2: oh no it doesn't! it scales the whole *model*, which is benign so long as all parameters of the dist are fixed. Not sure what the effect would be of re-scaling the whole model at every mcmc-step, let alone in ADVI, when the params are not fixed).

Then: should Bound still be a class/method of its own, or should it be deprecated in favour of Truncatable(-with-bound-fallback)?

**Comment by junpenglao** [comment]
> Then: should Bound still be a class/method of its own, or should it be deprecated in favour of Truncatable(-with-bound-fallback)?

That would be great. However, we dont have yet CDF implemented (there are some [WIP](https://github.com/pymc-devs/pymc3/pull/2688)) so it would be a bit difficult to implement for most distribution.

**Comment by gBokiau** [comment]
I don't think there's any downside here to laying the groundwork and filling the gaps on a rolling basis. Also, this requires CDF, not logCDF. That might all in all be more straight-forward to implement.

This also welcomes a `log(cdf(b)-cdf(a))` method which could, sometimes, be optimised.

File: pymc3/distributions/continuous.py [hunk 2]
```diff
@@ -434,6 +434,223 @@ def _repr_latex_(self, name=None, dist=None):
...
+            Desired size of random sample (returns one sample if not
+            specified).
+
+        Returns
+        -------
+        array
+        """
+        mu_v, std_v, a_v, b_v = draw_values([self.mu, self.sd, self.a, self.b], point=point, size=size)
+        return generate_samples(stats.truncnorm.rvs,
+                                                 a=(a_v - mu_v)/std_v,
```
Comment (twiecki) [inline pymc3/distributions/continuous.py:548]: pep8 and odd indentation.

Comment (junpenglao) [reply pymc3/distributions/continuous.py:548]: I will fix it in #3058

**Comment by fonnesbeck** [comment]
I was under the impression that Bound resulted in a valid PDF. I can check this, but if it does not, it should be removed.

**Comment by aseyboldt** [comment]
@fonnesbeck The pdf of `Bound` is not a proper distribution, it is not normalized. We actually say this in the docstring. In most cases that is not a problem, as we do not care about constants in the logp during mcmc sampling. I'm a bit torn here, as `Bound` is definitely useful sometimes (we don't need to worry about the cdf and ccdf functions), but it is also clearly a trap for users.

**Pull request merged**